pay/contracts/subscription/HectorSubscription.sol

```diff
@@ -0,0 +1,498 @@
+
+        // Pay first plan
+        if (balanceOf[msg.sender][plan.token] < plan.amount)
+            revert PAYER_IN_DEBT();
+
+        balanceOf[msg.sender][plan.token] -= plan.amount;
+        IERC20(plan.token).safeTransfer(treasury, plan.amount);
+
+        // Set subscription
+        Subscription storage subscription = subscriptions[msg.sender];
+        uint48 expiredAt = uint48(
+            max(block.timestamp, subscription.expiredAt)
+        ) + plan.period;
+
+        subscription.planId = _planId;
+        subscription.expiredAt = expiredAt;
+
+        emit SubscriptionCreated(msg.sender, _planId, expiredAt);
+    }
+
+    function depositAndCreateSubscription(uint256 _planId, uint256 _amount)
+        external
+    {
+        deposit(plans[_planId].token, _amount);
+        createSubscription(_planId);
+    }
+
+    function syncSubscriptions(address[] memory froms) external {
+        uint256 length = froms.length;
+        for (uint256 i = 0; i < length; i++) {
+            address from = froms[i];
+            if (subscriptions[from].planId > 0) {
+                syncSubscription(from);
+            }
+        }
+    }
+
+    function syncSubscription(address from) public {
+        Subscription storage subscription = subscriptions[from];
+        uint256 planId = subscription.planId;
+
+        if (planId == 0) revert INACTIVE_SUBSCRIPTION();
+
+        // before expiration
+        if (block.timestamp < subscription.expiredAt) {}
+        // after expiration
+        else {
+            Plan memory plan = plans[planId];
+            uint256 count = (block.timestamp -
+                subscription.expiredAt +
+                plan.period) / plan.period;
+            uint256 amount = plan.amount * count;
+            uint256 balance = balanceOf[from][plan.token];
+
+            // not active for now
+            if (balance < amount) {
+                count = balance / plan.amount;
+                amount = plan.amount * count;
+            }
+
+            if (count > 0) {
+                balanceOf[from][plan.token] -= amount;
+                subscription.expiredAt += uint48(plan.period * count);
+
+                IERC20(plan.token).transfer(treasury, amount);
+            }
+
+            // expired for a long time (deadline), then cancel it
+            if (subscription.expiredAt + expireDeadline <= block.timestamp) {
+                subscription.planId = 0;
+            }
+
+            emit SubscriptionSynced(
+                from,
+                subscription.planId,
+                subscription.expiredAt,
+                amount
+            );
+        }
+    }
+
+    function cancelSubscription() external {
+        syncSubscription(msg.sender);
+
+        uint256 planId = subscriptions[msg.sender].planId;
+
+        subscriptions[msg.sender].planId = 0;
+
+        emit SubscriptionCancelled(msg.sender, planId);
+    }
+
+    function modifySubscription(uint256 _newPlanId)
```

Comment:
@bejaxer will you create a view extracted from this func so that the dApp can inform users the following?
- We'll inform users the difference they have to pay in what token and/or the refund amount they expect to receive in their wallet. 
     - In the case of refund, return a boolean flag isRefund so that the dApp can store the following record to the DB so that support staff can process the refund request. If we didn't do this, there would be a communication gap between what the user expects (refund) and when the support staff can process it (they don't know). 
     - The following info will need to be returned from the view: 
         - Address (to)
         - Amount refund
         - Old plan
         - New plan

Reply:
using the callStatic as view func 
![image](https://user-images.githubusercontent.com/96207654/222180549-9bf3b8fc-b5ec-4316-bc54-ad172d9e666e.png)